logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/search/SearchUsersUseCase.kt

```diff
@@ -0,0 +1,123 @@
+
+/**
+ * Use case for searching users.
+ */
+class SearchUsersUseCase internal constructor(
+    private val searchUserRepository: SearchUserRepository,
+    private val selfUserId: UserId
+) {
+    suspend operator fun invoke(
+        searchQuery: String,
+        excludingMembersOfConversation: ConversationId?,
+        customDomain: String?
+    ): Result {
+        return if (searchQuery.isBlank()) {
+            Result(
+                connected = searchUserRepository.getKnownContacts(excludingMembersOfConversation).getOrElse(emptyList()),
+                notConnected = emptyList()
+            )
+        } else {
+            handleSearch(searchQuery, excludingMembersOfConversation, customDomain)
+        }
+    }
+
+    private suspend fun handleSearch(
+        searchQuery: String,
+        excludingConversation: ConversationId?,
+        customDomain: String?
+    ): Result = coroutineScope {
+        val remoteResultsDeferred = async {
+            searchUserRepository.searchUserRemoteDirectory(
+                searchQuery,
+                customDomain ?: selfUserId.domain,
+                MAX_SEARCH_RESULTS,
+                SearchUsersOptions(
+                    conversationExcluded = excludingConversation?.let { ConversationMemberExcludedOptions.ConversationExcluded(it) }
+                        ?: ConversationMemberExcludedOptions.None,
+                    selfUserIncluded = false
+                )
+            )
+                .map { userSearchResult ->
+                    userSearchResult.result.map {
+                        UserSearchDetails(
+                            id = it.id,
+                            name = it.name,
+                            completeAssetId = it.completePicture,
+                            previewAssetId = it.previewPicture,
+                            type = it.userType,
+                            connectionStatus = it.connectionStatus
+                        )
+                    }
+                }.getOrElse(emptyList())
+                .associateBy { it.id }
+                .toMutableMap()
+        }
+
+        val localSearchResultDeferred = async {
+            searchUserRepository.searchLocalByName(searchQuery, excludingConversation)
+                .getOrElse(emptyList())
+                .associateBy { it.id }
+                .toMutableMap()
+        }
+
+        val remoteResults = remoteResultsDeferred.await()
+        val localSearchResult = localSearchResultDeferred.await()
+
+        // a list of updated user ids so it can be deleted from the remote results
+        val updatedUser = mutableListOf<UserId>()
+
+        remoteResults.forEach { (userId, remoteUser) ->
+            if (localSearchResult.contains(userId)) {
+                localSearchResult[userId] = remoteUser
+                updatedUser.add(userId)
+            }
+        }
+
+        updatedUser.forEach { userId ->
+            remoteResults.remove(userId)
+        }
```

Comment:
Suggestion: Maybe this could be extracted to a function with a name, so later we might want to remove the code comment `//`. 

Something like, `fun resolveLocalAndRemoteResults(remoteResults, updatedUser )` that returns the  
```
Result(connected = localSearchResult.values.toList(), notConnected = remoteResults.values.toList())
```